File: World Finals/costly-binary-search.cpp
// Copyright (c) 2015 kamyu. All rights reserved.

/*
 * Google Code Jam 2015 World Finals - Problem A. Costly Binary Search
 * https://code.google.com/codejam/contest/5224486/dashboard#s=p0
 *
 * Time:  O(NlogN)
 * Space: O(NlogN)
 *
 */

#include <iostream>
#include <string>
#include <algorithm>
#include <utility>

using std::cin;
using std::cout;
using std::endl;
using std::string;
using std::max;
using std::min;

// 1 <= N <= 10^6
const int MAX_N = 1e6;

// O(log2(N) * max(cost)) ~= 20 * 9 = 180
const int MAX_TIMES = 180;

// left[t][p]: the farest position left to p which forms the range
//             [left[t][p], p) we can cover in time up to t
int left[MAX_TIMES + 1][MAX_N + 1];

// right[t][p]: the farest position right to p which forms the range
//              [p, right[t][p]) we can cover in time up to t
int right[MAX_TIMES + 1][MAX_N + 1];

int costly_binary_search() {
    string S;
    cin >> S;
    const int n = S.length();
    for (int p = 0; p <= n; ++p) {
        left[0][p] = p;
        right[0][p] = p;
    }

    // Given t time, check if there is a point where
    // we'll be able to produce the entire array.
    for (int t = 1; t <= MAX_TIMES; ++t) {
        for (int p = 0; p <= n; ++p) {
            left[t][p] = left[t - 1][p];
            right[t][p] = right[t - 1][p];
        }
        for (int p = 0; p < n; ++p) {
            const int tt = t - (S[p] - '0');
            if (tt >= 0) {
                const int l = left[tt][p];
                const int r = right[tt][p + 1];
                // [l---]p[------r]
                left[t][r] = min(left[t][r], l);
                right[t][l] = max(right[t][l], r);
            }
        }

        for (int p = n; p > 0; --p) {
            // [lp   [l(p-1)      ](p-1)]p
            left[t][p - 1] = min(left[t][p - 1], left[t][p]);
        }
        for (int p = 0; p < n; ++p) {
            // [p[(p+1)    ][rp     ]r(p+1)
            right[t][p + 1] = max(right[t][p + 1], right[t][p]);
        }

        if (right[t][0] == n) {
            // Found the min time which covers the entire array.
            return t;
        }
    }
    return MAX_TIMES;
}

int main() {
    int T;
    cin >> T;

    for (int test = 1; test <= T; ++test) {
        cout << "Case #" << test << ": "
             << costly_binary_search() << endl;
    }
    return 0;
}

File: World Finals/costly-binary-search.py
# Copyright (c) 2015 kamyu. All rights reserved.
#
# Google Code Jam 2015 World Finals - Problem A. Costly Binary Search
# https://code.google.com/codejam/contest/5224486/dashboard#s=p0
#
# Time:  O(NlogN), pass in PyPy2 within 5 minutes (TLE for Python2), and C++ within 2 minutes
# Space: O(N)
#

# 1 <= N <= 10^6
MAX_N = 10 ** 6

# O(log2(N) * max(cost)) ~= 20 * 9 = 180
MAX_TIMES = 180

# left[t][p]: the farest position left to p which forms the range
#             [left[t][p], p) we can cover in time up to t
left = [[0 for _ in xrange(MAX_N + 1)] for _ in xrange(10)]

# right[t][p]: the farest position right to p which forms the range
#              [p, right[t][p]) we can cover in time up to t
right = [[0 for _ in xrange(MAX_N + 1)] for _ in xrange(10)]

def costly_binary_search():
    S = raw_input().strip()
    n = len(S)
    for p in xrange(n + 1):
        left[0][p] = p
        right[0][p] = p

    # Given t time, check if there is a point where
    # we'll be able to produce the entire array.
    for t in xrange(1, MAX_TIMES + 1):
        for p in xrange(n + 1):
            left[t % 10][p] = left[(t - 1) % 10][p]
            right[t % 10][p] = right[(t - 1) % 10][p]
        for p in xrange(n):
            tt = t - int(S[p])
            if tt >= 0:
                l = left[tt % 10][p]
                r = right[tt % 10][p + 1]
                # --(t-S[p])--   --(t-S[p])--
                # [l          ]p[           r]
                left[t % 10][r] = min(left[t % 10][r], l)
                right[t % 10][l] = max(right[t % 10][l], r)

        for p in reversed(xrange(1, n + 1)):
            # ------------t------------
            #       ------t------
            # ---------t---------
            # [lp   [l(p-1)      ](p-1)]p
            left[t % 10][p - 1] = min(left[t % 10][p - 1], left[t % 10][p])
        for p in xrange(n):
            # -------t-------------
            #   -----t----
            #   ----------t--------
            # [p[(p+1)    ][rp     ]]r(p+1)
            right[t % 10][p + 1] = max(right[t % 10][p + 1], right[t % 10][p])

        if right[t % 10][0] == n:
            # Found the min time which covers the entire array.
            return t

    return MAX_TIMES


for case in xrange(input()):
    print "Case #%d: %d" % (case+1, costly_binary_search())
